src/aspire/image/__init__.py

```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+        if data.ndim == 2:
+            data = data[:, :, np.newaxis]
 
-    def __getitem__(self, item):
-        return self._data[item]
+        self.data = data.copy()
+        self.shape = self.data.shape
+        self.n_images = self.shape[-1]
+        self.res = self.shape[0]
 
-    def shift(self):
-        raise NotImplementedError
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __add__(self, other):
+        return Image(self.data + other.data)
+
+    def __repr__(self):
+        return f'{self.n_images} images of size {self.res}x{self.res}'
+
+    def asnumpy(self):
+        warnings.warn('Access to underlying Image data is discouraged.'
+                      'If possible, Image class should directly support called operation.')
+        return self.data
+
+    def shift(self, shifts):
+        """
+        Translate image by shifts. This method returns a new Image.
+        :param shifts: An array of size n-by-2 specifying the shifts in pixels.
+            Alternatively, it can be a column vector of length 2, in which case the same shifts is applied to each image.
+        :return: The Image translated by the shifts, with periodic boundaries.
+        """
+        if shifts.ndim == 1:
+            shifts = shifts[np.newaxis, :]
+
+        warnings.warn('Multiple implementations of im_translate. Running both for sanity check.'
+                      'Eliminating this will result in substantial speedup.')
+
+        im_translated = _im_translate(self.data, shifts)       # slower
+        im_translated2 = _im_translate2(self.data, -shifts.T)  # faster
+        assert np.allclose(im_translated, im_translated2)
+
+        return Image(im_translated)
+
+    def downsample(self, res):
+        """
+        Downsample Image to a specific resolution. This method returns a new Image.
+        :param res: int - new resolution, should be <= the current resolution of this Image
+        :return: The down-sampled Image object.
+        """
+        grid = grid_2d(self.res)
+        grid_ds = grid_2d(res)
+
+        im_ds = np.zeros((res, res, self.n_images)).astype(self.data.dtype)
+
+        # x, y values corresponding to 'grid'. This is what scipy interpolator needs to function.
+        res_by_2 = self.res / 2
+        x = y = np.ceil(np.arange(-res_by_2, res_by_2)) / res_by_2
+
+        mask = (np.abs(grid['x']) < res / self.res) & (np.abs(grid['y']) < res / self.res)
+        im = np.real(centered_ifft2(centered_fft2(self.data) * np.expand_dims(mask, 2)))
+
+        for s in range(im_ds.shape[-1]):
+            interpolator = RegularGridInterpolator(
+                (x, y),
+                im[:, :, s],
+                bounds_error=False,
+                fill_value=0
+            )
+            im_ds[:, :, s] = interpolator(np.dstack([grid_ds['x'], grid_ds['y']]))
+
+        return Image(im_ds)
+
+    def filter(self, filter=None, filter_values=None):
```

Comment:
I brought this up with Junchao last week.
https://github.com/ComputationalCryoEM/ASPIRE-Python/pull/32#discussion_r314874329

Not sure I understood why he needed this. Will bring this up next week.

Reply:
The computing is less if we pass the filters as precomputed values in 2D matrix since the number of filter types should be much smaller as the number of particle images.

Reply:
If the caller (`denoise_covar2d.py and the corresponding test script) simply used `sim.eval_filters` by passing in an Image object (in your case, the clean images), instead of first using `sim.eval_filter_grid` (which should be removed), then `eval_filters` groups processing by unique filters anyway, and calls the above `filter` method without having to worry about the `filter_values` parameter.

Reply:
Fixed in e85a5619bd1ea7bd8f5cac427b61f8f881f00e24. A separate issue can be opened in case this is unsatisfactory and there's something I'm missing.